lib/protocol/protocol.go

```diff
@@ -1078,3 +1057,51 @@ func (c *rawConnection) lz4Decompress(src []byte) ([]byte, error) {
 	}
 	return decoded, nil
 }
+
+type unwrapError interface {
+	error
+	Unwrap() error
+}
+
+type ProtocolError struct {
+	unwrapError
+}
+
+func newProtocolError(err error, msgContext string) *ProtocolError {
+	return &ProtocolError{
+		unwrapError: fmt.Errorf("protocol error on %v: %w", msgContext, err).(unwrapError),
+	}
+}
```

Comment:
```suggestion
func protocolError(err error, msgContext string) error {
	return fmt.Errorf("protocol error on %v: %w", msgContext, err)
}
```

I mean, at this point we don't use the types for formatting or recognition, so we could skip having them :)